### PR TITLE
#2098: runtime: add code to generate node IDs

### LINKED ISSUE #2098: Add physical node to LB data file
**What Needs to be Done?**

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #2098

### CODE REVIEW

**Comment by lifflander** [comment]
I think this a much better approach than `MPI_Get_processor_name`. It does rely on the MPI 3 standard.

**Review by PhilMiller — COMMENTED** [review]
This looks decent, though it would still be nice to have something like the host name, to be able to identify slow nodes

File: src/vt/runtime/runtime.cc [hunk 1]
```diff
@@ -201,6 +201,44 @@ Runtime::Runtime(
...
+
+  int num_nodes = -1;
+  int is_rank_0 = (shm_rank == 0) ? 1 : 0;
+  MPI_Allreduce(&is_rank_0, &num_nodes, 1, MPI_INT, MPI_SUM, i_comm);
+
+  int starting_rank = -1;
+  MPI_Comm_rank(i_comm, &starting_rank);
+
+  MPI_Comm node_number_comm;
+  MPI_Comm_split(i_comm, shm_rank, starting_rank, &node_number_comm);
```
[the remark this answers is not in the document]
Comment (lifflander) [reply src/vt/runtime/runtime.cc:228]: No, there is no assumption of equal numbers of ranks per physical node. Basically I number each shared node from 0...s then I split another comm with all the rank 0's as a color. Each one of those gets it's rank, which will be a unique node. I broadcast that node ID to each rank in the shared node.

**Review by PhilMiller — APPROVED** [review]

File: src/vt/runtime/runtime.cc
Comment (nlslatt) [inline src/vt/runtime/runtime.cc:1]: Looks great!

**Comment by nlslatt** [comment]
@lifflander I can confirm that this worked in the following configurations:

(a) 48 ranks per node using 8 nodes, where the first node gets ranks 0-47, the second gets ranks 48-95, etc.;
(b) similar to (a) but where I limit the total number of processes so that there are only 48*8-2 of them (which puts only 46 on node 7); and
(c) similar to (b) but using `--bynode`, where the ranks get assigned round robin except that node 7 still only gets 46 ranks (confirmed by sshing into node 7 and counting the running processes).

File: src/vt/vrt/collection/balance/node_lb_data.cc [hunk 2]
```diff
@@ -194,8 +194,25 @@ void NodeLBData::createLBDataFile() {
...
   if (not lb_data_writer_) {
+    auto prefix = fmt::format(
+      "\"type\":\"{}\","
+      "\"rank\":\"{}\","
+      "\"shared_node\":{{"
+        "\"id\":{},"
+        "\"size\":{},"
+        "\"rank\":{},"
+        "\"num_nodes\":{}"
+      "}},",
```
Comment (nlslatt) [inline src/vt/vrt/collection/balance/node_lb_data.cc:205]: Should we use the json utility to write this instead of hard-coding it?

**Review by nlslatt — COMMENTED** [review]
The schema validator is failing in one of the pipelines. The switch from a simple string schema type to using hard-coded json snippets in several places (including tests) feels hackish. Would it really be that hard to fix the interface so that we can use the json utility to write all the json?

**Comment by lifflander** [comment]
Looks great to me!

**Comment by ppebay** [comment]
As a side note, this branch's JSON schema validator is passing for both the "toy user defined" problem:
```
[LBAF_app] Executing LBAF version 0.1.0rc1
[LBAF_app] Executing with Python 3.8.16
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.0.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.1.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.2.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.3.json VT object map
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.1.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.2.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.3.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/user-defined-memory-toy-problem/toy_mem.0.json
```
as well as for the "challenging problem with fewer tasks":
```
[LBAF_app] Executing LBAF version 0.1.0rc1
[LBAF_app] Executing with Python 3.8.16
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.0.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.1.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.2.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.3.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.4.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.5.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.6.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.7.json VT object map
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.0.json
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.8.json VT object map
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.2.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.1.json
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.9.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.10.json VT object map
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.3.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.7.json
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.11.json VT object map
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.6.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.5.json
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.12.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.13.json VT object map
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.4.json
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.14.json VT object map
[lbsVTDataReader] Reading /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.15.json VT object map
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.8.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.11.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.10.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.9.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.13.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.15.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.12.json
[lbsVTDataReader] Valid JSON schema in /Users/pppebay/Documents/Git/LB-analysis-framework/data/challenging_toy_fewer_tasks/toy.14.json
```

I would like to suggest @nlslatt that the version of the validator also be reported in the logger output

File: src/vt/runtime/runtime.h [hunk 3]
```diff
@@ -430,6 +435,12 @@ struct Runtime {
 
   static bool volatile sig_user_1_;
 
+  bool has_physical_node_info = false;
+  int physical_node_id = -1;
+  int physical_num_nodes = -1;
+  int physical_node_size = -1;
+  int physical_node_rank = -1;
```
Comment (PhilMiller) [inline src/vt/runtime/runtime.h:442]: Should these all be suffixed with underscores?

**Review by PhilMiller — APPROVED** [review]
I'm happy with the node identification code. Maybe Philippe should review the json bits.

**Comment by nlslatt** [comment]
> I would like to suggest @nlslatt that the version of the validator also be reported in the logger output

As I don't think we have any version number for this right now, let's defer this to another issue.

**Review by nlslatt — APPROVED** [review]
I confirmed that the code to analyze the physical nodes works and that the resulting JSON is as expected.

**Pull request merged**